옥정현/refactor 41 jwt 재발급 로직 수정, requestBody 누락 예외 처리 구현

okjunghyeon opened the pull request:
## PR 유형  
어떤 변경 사항이 있나요?  

- [x] 새로운 기능 추가  
- [ ] 버그 수정  
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)  
- [x] 코드 리팩토링(성능, 기능 메서드)  
- [ ] 주석 추가 및 수정  
- [ ] 문서 수정  
- [x] 테스트 추가, 테스트 리팩토링  
- [ ] 빌드 부분 혹은 패키지 매니저 수정  
- [ ] 파일 혹은 폴더명 수정  
- [ ] 파일 혹은 폴더 삭제  

---

## 구현 내용  

### ✅ jwt 재발급 로직 수정
##### CustomAuthenticationFilter의 refreshAccessToken 메서드
- 토큰 재발급 과정에 AccessToken의 payload를 확인해 바로 User를 반환해주는 로직을 제거하였습니다. 따라서 메서드명에 맞게 refreshToken을 통한 검증 및 재발급만 가능하도록 수정되었습니다.
- 로그아웃 등의 이유로 refreshToken이 만료되었으나 AccessToken은 정상작동하여 인증이 통과되던 문제를 해결했습니다.




### ✅ requestForm 클래스들을 Dto 패키지에서 Controller 내부로 이동
- controller에서 request를 받는 목적으로만 사용되는 Form record 클래스를 Dto 패키지에서 controller 내부로 위치했습니다.


### ✅requestBody 누락에 대한 예외 처리 구현
- api 호출 시 requestBody를 누락할 경우 예외를 발생하도록 ExceptionHandler를 추가하였습니다.
- 누락 예외 상황에 대한 테스트 케이스를 추가하였습니다.


#### ➕ 커밋 메시지 컨벤션에 맞게 controller, service 메서드를 수정했습니다.

---

## PR Checklist  
PR이 다음 요구 사항을 충족하는지 확인하세요.  

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 진행했습니다. (버그 수정 / 기능에 대한 테스트)  

---

## ✅ **PR 포인트 & 검토 요청 사항**

okjunghyeon committed:
refactor: refreshAccessToken 메서드에서 AccessToken으로 User를 찾는 과정 제거

- 토큰 refresh 과정에서 AccessToken으로 다시 User를 반환하는 불필요한 과정 제거
- refreshToken이 만료되었으나 AccessToken이 정상 작동하여 인증이 통과되는 예외 상황 방지

okjunghyeon committed:
feat: 로그아웃 시 사용자 인증정보를 제거하도록 setLogout 기능 구현

okjunghyeon committed:
refactor: CRUD method 명명 규칙에 맞게 메서드명 수정

okjunghyeon committed:
feat: handleHttpMessageNotReadableException Handler를 추가하여 RequestBody 누락에 대한 검증 및 테스트 케이스 추가

okjunghyeon committed:
refactor: Usercontroller 내부에서만 사용되는 RequestForm들을 Dto 패키지에서 Controller 내부로 이동

- loginUserForm의 경우 처리해야하는 validation이 많아 Dto 패키지에 유지

omegafrog commented:
customFilter에서 다른 exception들을 먹어버리는 것 같습니다. 이게 Exception이 발생해서 handler에서 걸러지지 않는 녀석들이 나오면 /error로 spring이 자동으로 리다이렉트합니다. 그런데 이때 /error에 인증 정보가 없어서 401에러가 나고 결국 404나 400 에러가 나야 하는데 그것들이 다 400 에러로 바뀌는 것 같습니다.

그래서 customFilter에 `/error`를 인증정보를 검사하지 않도록 통과시켜야 할 것 같습니다

asungkim commented on backend/src/main/java/com/NBE_4_5_2/Team5/global/Rq.java:
```diff
@@ -36,6 +36,10 @@ public void setLogin(User actor) {
+        SecurityContextHolder.clearContext();
+    }
+
     public User getUserIdentity() {
```
제가 rq를 사용하다보니 getUserIdentity에서 토큰없이 정보를 접근하면 아래 401-3 쪽으로만 빠지고 로그인이 필요합니다는 안나오길래 확인해봤는데  인증되지 않은 사용자가 접근하면 authentication 이 null 이 안되고 Anonymous로 되더라고요. 
그래서 authentication 이 AnonymousToken인지 체크를 해봐야될거같아요.

okjunghyeon replied:
네 알겠습니다 CustomAuthenticationFilter에 검증 과정 추가하도록 하겠습니다!

okjunghyeon replied:
반영하고 테스트 케이스 추가했습니다
확인하시고 approve 해주시면 merge하겠습니다

asungkim reviewed (changes requested):
if 문에 널 체크외에도 체크를 추가해야할거같아요

okjunghyeon committed:
fix: 인증이 필요한 기능 사용 시 사용자 인증정보가 null이 아닌 Anonymous로 넘어오는 경우 예외 처리

- 토큰이 없으나 인증이 되어있는 경우, anonymous 혹은 인증되지 않은 사용자가 접근했을 때에 대한 처리와 테스트케이스 작성
- 쿠키 혹은 헤더로 인증하는 상황에 대한 테스트 케이스 추가

okjunghyeon committed:
fix: "/error" 경로를 requestMatchers에 추가하여 예외 발생 상황에 인증이 걸리지 않도록 수정

- 인증된 사용자가 api 이용 중 Exceptionhandler에서 잡히지 않는 Exception을 접했을 때 실제 예외가 아닌 401 인증 필요 예외가 발생하지 않도록 수정
- 인증이 열려있지 않는 api 혹은 인증 되지 않은 사용자의 경우 동일하게 401 인증 필요 예외 발생
- 테스트 케이스 추가

okjunghyeon committed:
merge conflict resolve, 옥정현/refactor-41 into develop

okjunghyeon committed:
refactor: 수정된 코드에 맞게 import 설정 및 yml 추가

omegafrog commented:
```
                        .requestMatchers("/h2-console/**")
                        .permitAll()
                        .requestMatchers("/api/posts/**", "/api/comments/**", "/api/payments/**", "/api/admin/**", "/api/chat/**", "/api/chatting/**")
                        .permitAll()
                        .requestMatchers("/api/users/login", "/api/users/signup", "/api/users/refresh")
                        .permitAll()
                        .anyRequest()
                        .authenticated()
```
이렇게 설정이 되었는데 기본적으로 모든 요청을 막고 저희가 만든 API는 일단 다 뚫려있는 상태입니다.

만약에 이때 "/api/users/logins" 같이 저희가 설정해두지 않은 요청을 보낸다면  404에러가 나와야 되는데 인증정보가 없기 때문에 401에러가 나오게 됩니다.
![image](https://github.com/user-attachments/assets/53370360-2989-46e0-8ecb-d3fec7ce7013)
( POST `/api/users/logins` 로 요청을 보냈는데 401 에러가 나오는 모습)

그래서 모든 요청을 열어두고, 저희가 만든 API 중에서 필요한 것만 막도록 해야 할 것 같습니다.

해당 문제는 이슈로 만들어 올리겠습니다

okjunghyeon commented:
네 이해했습니다. 모든 주소를 열어두고 인증이 필요한 api인 경우
Rq 클래스의 getUserIdentity 메서드를 통해 authentication을 검증하는 방식으로 사용하면 될 것 같습니다.
